File: Projects/MultiFileReader/modules/xml_parser.py
#
# Nimish Nayak
# 08-06-2017
#

#
# XML Parser 
#

# required libraries
import logging
import xml.dom.minidom
from parser import Parser
from utils import Utils

class XmlParser(Parser):

    # initialize the instance variables
    def __init__(self, input_file):
    	# call the base class constructor
    	Parser.__init__(self, input_file)
        logging.debug("Xml parser initialized")

    # parse the file using a xml parser 
    def load_file(self):
        # use the parse() function to load and parse an XML file
        try:
            self.doc = xml.dom.minidom.parse(self.input_file)
        except xml.parsers.expat.ExpatError as e:
            logging.info(e)
            self.doc = None

    # parse the file using a xml parser 
    def parse_file(self):
        # get a list of XML tags from the document and create the list of dictionaries 
        if self.doc:
            users = self.doc.getElementsByTagName("user")
            logging.debug("%d users:" % users.length)
            
            for user in users:    
                
                # Default Value
                val = {"name":"","value":0,"active":False}

                # get all the tag values 
                name = user.getElementsByTagName("name")    
                value = user.getElementsByTagName("value")    
                active = user.getElementsByTagName("active")    

                if name and name[0] and name[0].firstChild:
                    val["name"] = str(name[0].firstChild.nodeValue)

                if active and active[0] and active[0].firstChild:
                    val["active"] = Utils.str_to_bool(active[0].firstChild.nodeValue)
                
                if value and value[0] and value[0].firstChild and val["active"]:
                    val["value"] = int(value[0].firstChild.nodeValue)

                self.data.append(val)
            logging.debug("data saved: %s"%self.data)

File: modules/xml_parser.php
<?php

#
# Nimish Nayak
# 10-06-2017
#

#
# XML Parser 
#

# Note: To get the XML Module - 

# Since PHP 7 is not in the official Ubuntu PPAs, installed it through Ondřej Surý's PPA (sudo add-apt-repository ppa:ondrej/php). 

# uncomment the following line in /etc/php/7.0/fpm/php.ini
# extension=php_xmlrpc.dll

# and install php7.0-xml:
# sudo apt-get install php7.0-xml

# restart PHP:
# sudo service php7.0-fpm restart

# required libraries
require_once("./modules/utils.php");
require_once("./modules/parser.php");

class XmlParser extends Parser {

    # initialize the instance variables
    function __construct($inputFile){
        
        # call the base class constructor
        parent::__construct($inputFile);
        echo("Xml parser initialized\n");
    }

    # load the file using an xml parser
    private function loadFile(){

        # read the whole file into a string
        $myXMLData = file_get_contents($this->inputFile);

        # string -> xml
        $this->doc = simplexml_load_string($myXMLData);
        
        # Debug 
        # var_dump($this->doc);

        if ($this->doc === false) {
            echo "Failed loading XML:\n";
            foreach(libxml_get_errors() as $error) {
                echo $error->message, "\n";
            }
        }
    }
        
    # parse the file using a xml parser 
    public function parseFile() {

        $this->loadFile();

        if ($this->doc->user) {

            # we have xml data!
            # parse and create an associative array from it
            $this->data = array();

            foreach ($this->doc->user as $user) {

                $dict = array('name' => $user->name, 'active' => Utils::strToBool($user->active), 'value' => (int)$user->value);
                array_push($this->data, $dict);

            }

        }

    }
}

?>